hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/opt/physical/SortPhysicalRule.java

```diff
@@ -148,18 +148,22 @@ private static boolean requiresLocalSort(RelCollation sortCollation, RelCollatio
         }
     }
 
-    private static SortPhysicalRel createLocalSort(SortLogicalRel logicalSort, RelNode physicalInput, boolean requiresLocalSort) {
+    private static SortPhysicalRel createLocalSort(SortLogicalRel logicalSort, RelNode physicalInput, boolean requiresLocalSort,
+                                                   boolean isFullResultOnAll) {
         // Input traits are propagated, but new collation is used.
         RelTraitSet traitSet = OptUtils.traitPlus(physicalInput.getTraitSet(),
             logicalSort.getCollation()
         );
 
+        // Don't push down the fetch/offset operators if merging phase is needed
         return new SortPhysicalRel(
             logicalSort.getCluster(),
             traitSet,
             physicalInput,
             logicalSort.getCollation(),
-            !requiresLocalSort
+            requiresLocalSort,
+            isFullResultOnAll ? logicalSort.offset : null,
```

Comment:
Pushing limit/offset here might produce invalid results because the order of rows is not defined. If there are two members, each of them may produce different results, which will lead to unpredictable results. Another reason to remove the unsorted limir/offset from the product.

Reply:
We're passing non-null values only if we have a full result on all members. For instance, if we have a one-member cluser. For other cases, we pass null doing the limit/offset on the sort-merge level. 
If the sorting is not required, the collation is null, and the `SortPhysicalRel` acts as a fetch operator. 
So, i'm not sure that I completely got the point about unpredictable results. Can you please elaborate more on this?

Reply:
`isFullResultOnAll` returns `true` if the relation is either `REPLICATED`, or there is only one member in the cluster. 
Consider that we have a `REPLICATED` relation. We do not have a guarantee that all members produce rows for this relation in the same order. E.g., on member produces `[a,b]`, whilst the other produces `[b,a]`. If we apply the `LIMIT 1`, we will get `[a]` on one member, and `[b]` on the other.

Reply:
I think that is totally fine if we run the query on various members we may get different results that are both valid. Even for sorting we don't support stable sorting and may get similar issues.

Reply:
This is not fine, because it will lead to incorrect results. Imagine, that we have [part_map JOIN replicated_rel]. Since the `replicated_rel` produces different results on different members, the result of the join will be incorrect. For example: 
```
Member 1
[1, a] [2, b] JOIN [a] -> [1, a]

Member 2:
[3, a] [4, b] JOIN [b] -> [4, b]
```

The combined result would be 
```
[1, a]
[4, b]
```

This result cannot be explained by any `LIMIT 1` that filtered either `[a]` or `[b]`.

Reply:
SELECT * from t LIMIT 1 may return any result  according to the LIMIT semantics. It doesn't guarantee any order and even repeatable results. It implies that if you do JOIN on top of LIMIT you may expect such sort of anomalies. To avoid this you have to add an explicit sorting or perform limit on top of JOIN.

Reply:
My example relates to the situation when `LIMIT 1` is used inside the query. E.g.:
`SELECT * FROM partitioned JOIN (SELECT * FROM replicated LIMIT 1) ON ...`

Reply:
That is not a problem of LIMIT, but a problem of JOIN implementation, that relies on unpredictable results of the underneath operators. To fix that in the future we may need to add some stable sorting on some fields. But again, that is not an issue of the LIMIT operator itself.

Reply:
As we discussed, I've disabled FETCH/OFFSET in nested SELECT. See https://github.com/hazelcast/hazelcast/pull/18091/commits/baa0f80f059bded1eaf14e31798ce4d084f5604f